File: hooks/prompt-router.py
#!/usr/bin/env python3
"""beforeSubmitPrompt hook: surface workflow-state and skill-routing hints.

Reads the user prompt, scans for repo-owned skill, agent, and phase keywords,
and when an active workflow-state document is reachable (via the
`OH_MY_CURSOR_WORKFLOW_STATE` environment variable or a `workflow_state` field
inside the prompt-submit event) summarizes the current phase and pending
acceptance criteria.

Behavior is intentionally bounded:

- the hook is **fail-open**; it always exits 0 and never blocks a prompt;
- it only emits structured JSON; Cursor decides whether to surface the
  `additional_context` string;
- it does **not** mutate workflow state; it only reads it.
"""
from __future__ import annotations

import json
import os
import re
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from _trace import trace as _trace  # noqa: E402

# Resolve ROOT from environment or workspace, not __file__ (which may be unreliable in hook contexts)
ROOT = Path(os.environ.get("OH_MY_CURSOR_WORKSPACE", os.getcwd())).resolve()
if not (ROOT / "hooks").exists() or "plugins/local/oh-my-cursor" in str(ROOT):
    ROOT = Path(__file__).resolve().parents[1]

SKILL_NAMES = {
    "phase-controller", "plan", "auto-execute", "iterate-loop",
    "parallel-batch", "review", "debug", "trace", "security-review",
    "deep-interview", "doctor", "local-plugin-check", "verify", "mcp-setup",
}

AGENT_NAMES = {
    "orchestrator", "researcher", "planner", "implementer", "verifier",
    "critic", "debugger", "security-reviewer",
    "explore", "code-reviewer", "test-engineer", "tracer",
}

PHASE_NAMES = {
    "intake", "research", "plan", "execute",
    "verify", "review", "done", "blocked",
}

PROMPT_FIELDS = ("prompt", "text", "userPrompt", "user_prompt", "message", "content")


def _read_payload() -> dict:
    raw = sys.stdin.read()
    if not raw.strip():
        return {}
    try:
        value = json.loads(raw)
    except json.JSONDecodeError:
        return {"_invalid_json": True}
    return value if isinstance(value, dict) else {"payload": value}


def _extract_prompt(payload: dict) -> str:
    for key in PROMPT_FIELDS:
        value = payload.get(key)
        if isinstance(value, str) and value.strip():
            return value
    nested = payload.get("input")
    if isinstance(nested, dict):
        for key in PROMPT_FIELDS:
            value = nested.get(key)
            if isinstance(value, str) and value.strip():
                return value
    return ""


def _matches(prompt_lower: str, names: set[str]) -> list[str]:
    found: list[str] = []
    for name in names:
        token = re.escape(name)
        if re.search(rf"(?<![A-Za-z0-9_-]){token}(?![A-Za-z0-9_-])", prompt_lower):
            found.append(name)
    return sorted(found)


def _resolve_state_path(payload: dict) -> Path | None:
    candidates: list[str] = []
    env_path = os.environ.get("OH_MY_CURSOR_WORKFLOW_STATE")
    if env_path:
        candidates.append(env_path)
    payload_path = payload.get("workflow_state") if isinstance(payload, dict) else None
    if isinstance(payload_path, str):
        candidates.append(payload_path)
    for raw in candidates:
        try:
            candidate = Path(raw).expanduser()
            if not candidate.is_absolute():
                candidate = (ROOT / candidate).resolve()
            if candidate.is_file():
                return candidate
        except (OSError, RuntimeError, ValueError):
            continue
    return None


def _summarize_state(state_path: Path) -> dict:
    try:
        data = json.loads(state_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return {"loaded": False}
    if not isinstance(data, dict):
        return {"loaded": False}
    pending: list[str] = []
    failed: list[str] = []
    criteria = data.get("acceptance_criteria")
    if isinstance(criteria, list):
        for item in criteria:
            if not isinstance(item, dict):
                continue
            status = str(item.get("status", "")).lower()
            label = str(item.get("id") or item.get("criterion") or "").strip()
            if not label:
                continue
            if status == "failed":
                failed.append(label)
            elif status not in {"passed", "skipped"}:
                pending.append(label)
    return {
        "loaded": True,
        "task_id": data.get("task_id"),
        "phase": data.get("phase"),
        "status": data.get("status"),
        "current_role": data.get("current_role"),
        "next_action": data.get("next_action"),
        "pending_criteria": pending[:10],
        "failed_criteria": failed[:10],
    }


def main() -> int:
    payload = _read_payload()
    if payload.get("_invalid_json"):
        print(json.dumps({
            "status": "pass",
            "fail_open": True,
            "additional_context": "",
            "message": "Prompt-router input was not JSON; skipped.",
        }))
        return 0

    prompt = _extract_prompt(payload)
    prompt_lower = prompt.lower()

    matched_skills = _matches(prompt_lower, SKILL_NAMES)
    matched_agents = _matches(prompt_lower, AGENT_NAMES)
    matched_phases = _matches(prompt_lower, PHASE_NAMES)

    state_path = _resolve_state_path(payload)
    state_summary = _summarize_state(state_path) if state_path else {"loaded": False}

    parts: list[str] = []
    if matched_skills:
        parts.append(
            "Skill keywords detected: "
            + ", ".join(matched_skills)
            + ". Matching SKILL.md files live under skills/."
        )
    if matched_agents:
        parts.append(
            "Agent keywords detected: "
            + ", ".join(matched_agents)
            + ". Matching role prompts live under agents/."
        )
    if matched_phases:
        parts.append(
            "Phase keywords detected: "
            + ", ".join(matched_phases)
            + ". phase-controller routes phase transitions; see skills/phase-controller/SKILL.md."
        )
    if state_summary.get("loaded"):
        parts.append(
            "Active workflow state: phase="
            + str(state_summary.get("phase") or "?")
            + ", status="
            + str(state_summary.get("status") or "?")
            + (", role=" + str(state_summary.get("current_role")) if state_summary.get("current_role") else "")
            + "."
        )
        if state_summary.get("failed_criteria"):
            parts.append(
                "Failed acceptance criteria: "
                + ", ".join(state_summary["failed_criteria"])  # type: ignore[arg-type]
                + "."
            )
        if state_summary.get("pending_criteria"):
            parts.append(
                "Pending acceptance criteria: "
                + ", ".join(state_summary["pending_criteria"])  # type: ignore[arg-type]
                + "."
            )
        if state_summary.get("next_action"):
            parts.append("Recorded next action: " + str(state_summary["next_action"]) + ".")

    additional_context = " ".join(parts)
    output = {
        "status": "pass",
        "fail_open": True,
        "additional_context": additional_context,
        "matched_skills": matched_skills,
        "matched_agents": matched_agents,
        "matched_phases": matched_phases,
        "workflow_state": {
            "path": state_path.as_posix() if state_path else None,
            "loaded": state_summary.get("loaded", False),
            "phase": state_summary.get("phase"),
            "status": state_summary.get("status"),
            "current_role": state_summary.get("current_role"),
            "pending_criteria": state_summary.get("pending_criteria", []),
            "failed_criteria": state_summary.get("failed_criteria", []),
        },
    }
    _trace({
        "hook": "prompt-router",
        "event": "beforeSubmitPrompt",
        "status": output.get("status"),
        "prompt_length": len(prompt),
        "matched_skills": matched_skills,
        "matched_agents": matched_agents,
        "matched_phases": matched_phases,
        "state_loaded": state_summary.get("loaded", False),
    })
    print(json.dumps(output, ensure_ascii=False))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: hooks/prompt-router.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as process from 'node:process';
import { fileURLToPath } from 'node:url';
import { resolveWorkspaceRoot } from './_repo.ts';
import { trace } from './_trace.ts';

const currentFile = fileURLToPath(import.meta.url);
const WORKSPACE_ROOT = resolveWorkspaceRoot(currentFile);

export const SKILL_NAMES = new Set([
  'phase-controller', 'plan', 'auto-execute', 'iterate-loop',
  'parallel-batch', 'review', 'debug', 'trace', 'security-review',
  'deep-interview', 'doctor', 'local-plugin-check', 'verify', 'mcp-setup',
  'team-controller',
]);

export const AGENT_NAMES = new Set([
  'architect',
  'orchestrator', 'researcher', 'planner', 'implementer', 'verifier',
  'critic', 'debugger', 'security-reviewer',
  'explore', 'code-reviewer', 'qa-tester', 'test-engineer', 'tracer',
]);

export const PHASE_NAMES = new Set([
  'intake', 'research', 'plan', 'execute',
  'verify', 'review', 'done', 'blocked',
]);

const PROMPT_FIELDS = ['prompt', 'text', 'userPrompt', 'user_prompt', 'message', 'content'];

function readPayload(): any {
  try {
    const raw = fs.readFileSync(0, 'utf-8');
    if (!raw.trim()) {
      return {};
    }
    return JSON.parse(raw);
  } catch (err: any) {
    return { _invalid_json: true };
  }
}

function extractPrompt(payload: any): string {
  if (!payload || typeof payload !== 'object') {
    return '';
  }
  for (const key of PROMPT_FIELDS) {
    const value = payload[key];
    if (typeof value === 'string' && value.trim()) {
      return value;
    }
  }
  const nested = payload.input;
  if (nested && typeof nested === 'object') {
    for (const key of PROMPT_FIELDS) {
      const value = nested[key];
      if (typeof value === 'string' && value.trim()) {
        return value;
      }
    }
  }
  return '';
}

function matches(promptLower: string, names: Set<string>): string[] {
  const found: string[] = [];
  for (const name of names) {
    const escaped = name.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');
    const regex = new RegExp(`(?<![A-Za-z0-9_-])${escaped}(?![A-Za-z0-9_-])`, 'i');
    if (regex.test(promptLower)) {
      found.push(name);
    }
  }
  return found.sort();
}

function resolveStatePath(payload: any): string | null {
  const candidates: string[] = [];
  const envPath = process.env.OH_MY_CURSOR_WORKFLOW_STATE;
  if (envPath) {
    candidates.push(envPath);
  }
  const payloadPath = payload?.workflow_state;
  if (typeof payloadPath === 'string') {
    candidates.push(payloadPath);
  }

  for (const raw of candidates) {
    try {
      let candidate = raw;
      if (candidate.startsWith('~')) {
        const homedir = process.env.HOME || '';
        candidate = path.join(homedir, candidate.slice(1));
      }
      if (!path.isAbsolute(candidate)) {
        candidate = path.resolve(WORKSPACE_ROOT, candidate);
      }
      if (fs.existsSync(candidate) && fs.statSync(candidate).isFile()) {
        return candidate;
      }
    } catch {
      continue;
    }
  }
  return null;
}

function summarizeState(statePath: string): any {
  try {
    const data = JSON.parse(fs.readFileSync(statePath, 'utf-8'));
    if (!data || typeof data !== 'object') {
      return { loaded: false };
    }
    const pending: string[] = [];
    const failed: string[] = [];
    const criteria = data.acceptance_criteria;
    if (Array.isArray(criteria)) {
      for (const item of criteria) {
        if (!item || typeof item !== 'object') continue;
        const status = String(item.status || '').toLowerCase();
        const label = String(item.id || item.criterion || '').trim();
        if (!label) continue;
        if (status === 'failed') {
          failed.push(label);
        } else if (status !== 'passed' && status !== 'skipped') {
          pending.push(label);
        }
      }
    }
    return {
      loaded: true,
      task_id: data.task_id,
      phase: data.phase,
      status: data.status,
      current_role: data.current_role,
      next_action: data.next_action,
      pending_criteria: pending.slice(0, 10),
      failed_criteria: failed.slice(0, 10),
    };
  } catch {
    return { loaded: false };
  }
}

function main(): number {
  const payload = readPayload();
  if (payload?._invalid_json) {
    console.log(JSON.stringify({
      status: 'pass',
      fail_open: true,
      additional_context: '',
      message: 'Prompt-router input was not JSON; skipped.',
    }));
    return 0;
  }

  const prompt = extractPrompt(payload);
  const promptLower = prompt.toLowerCase();

  const matchedSkills = matches(promptLower, SKILL_NAMES);
  const matchedAgents = matches(promptLower, AGENT_NAMES);
  const matchedPhases = matches(promptLower, PHASE_NAMES);

  const statePath = resolveStatePath(payload);
  const stateSummary = statePath ? summarizeState(statePath) : { loaded: false };

  const parts: string[] = [];
  if (matchedSkills.length > 0) {
    parts.push(`Skill keywords detected: ${matchedSkills.join(', ')}. Matching SKILL.md files live under skills/.`);
  }
  if (matchedAgents.length > 0) {
    parts.push(`Agent keywords detected: ${matchedAgents.join(', ')}. Matching role prompts live under agents/.`);
  }
  if (matchedPhases.length > 0) {
    parts.push(`Phase keywords detected: ${matchedPhases.join(', ')}. phase-controller routes phase transitions; see skills/phase-controller/SKILL.md.`);
  }
  if (stateSummary.loaded) {
    parts.push(
      `Active workflow state: phase=${stateSummary.phase || '?'}, status=${stateSummary.status || '?'}` +
      (stateSummary.current_role ? `, role=${stateSummary.current_role}` : '') + '.'
    );
    if (stateSummary.failed_criteria && stateSummary.failed_criteria.length > 0) {
      parts.push(`Failed acceptance criteria: ${stateSummary.failed_criteria.join(', ')}.`);
    }
    if (stateSummary.pending_criteria && stateSummary.pending_criteria.length > 0) {
      parts.push(`Pending acceptance criteria: ${stateSummary.pending_criteria.join(', ')}.`);
    }
    if (stateSummary.next_action) {
      parts.push(`Recorded next action: ${stateSummary.next_action}.`);
    }
  }

  const additionalContext = parts.join(' ');
  const output = {
    status: 'pass',
    fail_open: true,
    additional_context: additionalContext,
    matched_skills: matchedSkills,
    matched_agents: matchedAgents,
    matched_phases: matchedPhases,
    workflow_state: {
      path: statePath,
      loaded: stateSummary.loaded,
      phase: stateSummary.phase || null,
      status: stateSummary.status || null,
      current_role: stateSummary.current_role || null,
      pending_criteria: stateSummary.pending_criteria || [],
      failed_criteria: stateSummary.failed_criteria || [],
    },
  };

  trace({
    hook: 'prompt-router',
    event: 'beforeSubmitPrompt',
    status: output.status,
    prompt_length: prompt.length,
    matched_skills: matchedSkills,
    matched_agents: matchedAgents,
    matched_phases: matchedPhases,
    state_loaded: stateSummary.loaded,
  });

  console.log(JSON.stringify(output));
  return 0;
}

process.exit(main());
